File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
    const N = maps.length;
    const M = maps[0].length;
    const result = [];
    const visited = Array.from(new Array(N), () => new Array(M).fill(false));
    let sum = 0;
    
    for (let r = 0; r < N; r++) {
        for (let c = 0; c < M; c++) {
            if (visited[r][c]) continue;
            if (maps[r][c] === 'X') continue;
            
            checkIsland(r, c);

            if (sum > 0) result.push(sum);
            
            sum = 0;
        }
    }
    
    return result.length > 0 ? result.sort((a, b) => a - b) : [-1];
    
    function checkIsland(r, c) {
        sum += +maps[r][c];
        visited[r][c] = true;
        
        const dr = [-1, 0, 1, 0];
        const dc = [0, 1, 0, -1];
        
        for (let i = 0; i < 4; i++) {
            const nr = r + dr[i];
            const nc = c + dc[i];
            
            if (nr < 0 || nr >= N) continue;
            if (nc < 0 || nc >= M) continue;
            if (maps[nr][nc] === 'X') continue;
            if (visited[nr][nc]) continue;
            
            checkIsland(nr, nc, sum);
        }
    }
}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
def solution(maps):
    N = len(maps)
    M = len(maps[0])
    visited = [[False for _ in range(M)] for _ in range(N)]
    di = [1, 0, -1, 0]
    dj = [0, 1, 0, -1]
    result = []
    
    for i in range(N):
        for j in range(M):
            if visited[i][j] or maps[i][j] == 'X':
                continue
                
            stack = [[i, j]]
            visited[i][j] = True
            food = 0
            
            while len(stack) > 0:
                ci, cj = stack.pop()
                food += int(maps[ci][cj])
                
                for k in range(4):
                    ni = ci + di[k]
                    nj = cj + dj[k]
                    
                    if (ni >= 0 and nj >= 0 and ni < N and nj < M
                        and visited[ni][nj] == False and maps[ni][nj].isdigit()):
                        visited[ni][nj] = True
                        stack.append([ni, nj])
            
            if food > 0:
                result.append(food)

    return sorted(result) if len(result) > 0 else [-1]